tests/infer/test_predictive.py

```diff
@@ -0,0 +1,92 @@
+
+
+def one_hot_model(pseudocounts, classes=None):
+    probs_prior = dist.Dirichlet(pseudocounts)
+    probs = pyro.sample("probs", probs_prior)
+    with pyro.plate("classes", classes.size(0) if classes is not None else 1, dim=-1):
+        return pyro.sample("obs", dist.OneHotCategorical(probs), obs=classes)
+
+
+def beta_guide(num_trials):
+    phi_c0 = pyro.param("phi_c0", num_trials.new_tensor(5.0).expand([num_trials.size(0)]))
+    phi_c1 = pyro.param("phi_c1", num_trials.new_tensor(5.0).expand([num_trials.size(0)]))
+    with pyro.plate("data", num_trials.size(0)):
+        phi_posterior = dist.Beta(concentration0=phi_c0, concentration1=phi_c1)
+        pyro.sample("phi", phi_posterior)
+
+
+def test_posterior_predictive_svi_manual_guide():
+    true_probs = torch.ones(5) * 0.7
+    num_trials = torch.ones(5) * 1000
+    num_success = dist.Binomial(num_trials, true_probs).sample()
+    conditioned_model = poutine.condition(model, data={"obs": num_success})
+    svi = SVI(conditioned_model, beta_guide, optim.Adam(dict(lr=1.0)), Trace_ELBO())
+    for i in range(1000):
+        svi.step(num_trials)
+    posterior_samples = predictive(beta_guide, {}, num_trials[:3], num_samples=100)
+    posterior_predictive = predictive(model, posterior_samples, num_trials[:3], num_samples=10000)
+    #     svi.step(num_trials[:3])
+    # posterior_samples = predictive(beta_guide, {}, num_trials, num_samples=100)
+    # posterior_predictive = predictive(model, posterior_samples, num_trials, num_samples=10000)
+    marginal_return_vals = posterior_predictive["_RETURN"]
+    assert_close(marginal_return_vals.mean, torch.ones(3) * 700, rtol=0.05)
+
+
+def test_posterior_predictive_svi_auto_delta_guide():
+    true_probs = torch.ones(5) * 0.7
+    num_trials = torch.ones(5) * 1000
+    num_success = dist.Binomial(num_trials, true_probs).sample()
+    conditioned_model = poutine.condition(model, data={"obs": num_success})
+    guide = AutoDelta(conditioned_model)
+    svi = SVI(conditioned_model, guide, optim.Adam(dict(lr=1.0)), Trace_ELBO())
+    for i in range(1000):
+        svi.step(num_trials)
+    posterior_samples = predictive(guide, {}, num_trials, num_samples=10000)
+    posterior_predictive = predictive(model, posterior_samples, num_trials)
+    marginal_return_vals = posterior_predictive["obs"]
+    assert_close(marginal_return_vals.mean(dim=0), torch.ones(5) * 700, rtol=0.05)
+
+
+def test_posterior_predictive_svi_auto_diag_normal_guide():
+    true_probs = torch.ones(5) * 0.7
+    num_trials = torch.ones(5) * 1000
+    num_success = dist.Binomial(num_trials, true_probs).sample()
+    conditioned_model = poutine.condition(model, data={"obs": num_success})
+    guide = AutoDiagonalNormal(conditioned_model)
+    svi = SVI(conditioned_model, guide, optim.Adam(dict(lr=1.0)), Trace_ELBO())
+    for i in range(1000):
+        svi.step(num_trials)
+    posterior_samples = predictive(guide, {}, num_trials, num_samples=10000)
```

Comment:
I think we should have `predictive` take in an optional `guide` keyword argument and do this behind the scenes, as @ahmadsalim suggested.

Reply:
I think that will also let us return `_RETURN` sites from the model trace.

Reply:
Yes, the behavior of having `guide` keyword should cover usage cases I can think about. I agree that we can support it for user convenience. I'll go for a final pass before requesting your review. :)